### PR TITLE
Implement rank  syndrome decoding estimator

### PR DESCRIPTION
**Pull request by ricardovp85**
### Description
-  This PR implements the rank syndrome decoding (SD) estimator

### Review process
For the rank syndrome decoding estimator, the following algorithms are implemented.
- Basis Enumeration (Combinatorial Algorithm)
- OJ 1  (Combinatorial algorithm)
-  OJ 2  (Combinatorial algorithm)
- GRS  (Combinatorial algorithm)
- Improved GRS  (Combinatorial algorithm)
- Guessing Enhanced GRS  (Combinatorial algorithm)
- Hybrid Linearization (Algebraic Algorithm)
- Max Minors (Algebraic Algorithm)
- Support Minors (Algebraic Algorithm)  

### Pre-approval checklist
- [x] The code builds clean without any errors or warnings
- [x] I've added/updated tests
- [x] I've added/updated documentation if needed

### CODE REVIEW

File: cryptographic_estimators/RSDEstimator/RSDAlgorithms/support_minors.py [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+class SupportMinors(RSDAlgorithm):
+    """
+    Construct an instance of SupportMinors estimator
+
+    This algorith is introduced in
+    M. Bardet, P. Briaud, M. Bros, P. Gaborit, and J.-P. Tillich,
+    “Revisiting algebraic attacks on MinRank and on the rank decoding problem",in DCC, 2023.
+
+     Args:
+         problem (MRProblem): An instance of the MRProblem class.
```
Comment (Memphisd) [inline cryptographic_estimators/RSDEstimator/RSDAlgorithms/support_minors.py:35]: .

File: cryptographic_estimators/RSDEstimator/RSDAlgorithms/support_minors.py [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+from math import comb as binomial
+
+
+class SupportMinors(RSDAlgorithm):
+    """
+    Construct an instance of SupportMinors estimator
+
+    This algorith is introduced in
+    M. Bardet, P. Briaud, M. Bros, P. Gaborit, and J.-P. Tillich,
+    “Revisiting algebraic attacks on MinRank and on the rank decoding problem",in DCC, 2023.
```
Comment (Memphisd) [inline cryptographic_estimators/RSDEstimator/RSDAlgorithms/support_minors.py:32]: .

File: cryptographic_estimators/RSDEstimator/rsd_algorithm.py [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+# ****************************************************************************
+
+
+from ..base_algorithm import BaseAlgorithm
+from .rsd_problem import RSDProblem
+
+
+class RSDAlgorithm(BaseAlgorithm):
+    def __init__(self, problem: RSDProblem, **kwargs):
+        """Base class for MR algorithms complexity estimator.
```
Comment (Memphisd) [inline cryptographic_estimators/RSDEstimator/rsd_algorithm.py:25]: MR algorithms

File: cryptographic_estimators/RSDEstimator/rsd_algorithm.py [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+            Args:
+                problem (MRProblem): MRProblem object including all necessary parameters
```
Comment (Memphisd) [inline cryptographic_estimators/RSDEstimator/rsd_algorithm.py:27]: .

File: cryptographic_estimators/RSDEstimator/rsd_problem.py [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+
+
+from ..base_problem import BaseProblem
+from .rsd_constants import *
+from math import log2, ceil
+from cryptographic_estimators.helper import is_prime_power, ngates
+
+
+class RSDProblem(BaseProblem):
+    """Construct an instance of RSDProblem.
```
Comment (Memphisd) [inline cryptographic_estimators/RSDEstimator/rsd_problem.py:26]: RankSDProblem

**Review by Memphisd — CHANGES_REQUESTED** [review]
Overall it looks good, just some smaller changes. 
Why are there changes to the Mayo and MQ estimator, couldn't follow that, is it just some code reusing re-structuring?

@Javierverbel @ricardovp85 Haven't checked the formulas yet, planning to study the overleaf soon.

@Memphisd  The Mayo and MQ estimator changes are related to code refactoring!

File: cryptographic_estimators/RSDEstimator/RSDAlgorithms/basis_enumeration.py [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+# ****************************************************************************
+
+
+from ..rsd_algorithm import RSDAlgorithm
```
[the remark this answers is not in the document]
Comment (ricardovp85) [reply cryptographic_estimators/RSDEstimator/RSDAlgorithms/basis_enumeration.py:19]: Thanks for your comment.  I'll change all the class names starting with RSD* to RankSD*

File: cryptographic_estimators/RSDEstimator/RSDAlgorithms/basis_enumeration.py [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+class BasisEnumeration(RSDAlgorithm):
+    """Construct an instance of Basis Enumerator estimator
+
+      This algorithm enumerates the possible supports for the vector x and is introduced in
+      Florent Chabaud and Jacques Stern,
+      "The Cryptographic Security of the Syndrome Decoding Problem for Rank Distance Codes", ASIACRYPT, 1996.
+
+
+       Args:
+            problem (MRProblem): An instance of the MRProblem class.
```
[the remark this answers is not in the document]
Comment (ricardovp85) [reply cryptographic_estimators/RSDEstimator/RSDAlgorithms/basis_enumeration.py:33]: The same here!

File: cryptographic_estimators/RSDEstimator/RSDAlgorithms/basis_enumeration.py [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+from ..rsd_problem import RSDProblem
+from math import log2
+
+
+class BasisEnumeration(RSDAlgorithm):
+    """Construct an instance of Basis Enumerator estimator
+
+      This algorithm enumerates the possible supports for the vector x and is introduced in
+      Florent Chabaud and Jacques Stern,
+      "The Cryptographic Security of the Syndrome Decoding Problem for Rank Distance Codes", ASIACRYPT, 1996.
```
[the remark this answers is not in the document]
Comment (ricardovp85) [reply cryptographic_estimators/RSDEstimator/RSDAlgorithms/basis_enumeration.py:29]: Ok, thanks!

File: cryptographic_estimators/RSDEstimator/RSDAlgorithms/guessing_enhanced_grs.py [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+from math import log2, ceil, floor
+
+
+class GuessingEnhancedGRS(RSDAlgorithm):
+    """
+    Construct an instance of GuessingEnhancedGRS estimator
+
+    This algorithm is introduced in
+    G. D’Alconzo2, A. Esser, A. Gangemi, and C. Sanna,
+    “Ryde with mira: Partial key exposure attacks on rank-based schemes”, 2024.
```
[the remark this answers is not in the document]
Comment (ricardovp85) [reply cryptographic_estimators/RSDEstimator/RSDAlgorithms/guessing_enhanced_grs.py:31]: I'll remove that reference,  since that paper is not public yet.

**Review by Javierverbel — CHANGES_REQUESTED** [review]

**Comment by Javierverbel** [comment]
The doc format should be 

```   
"""Construct an instance of RankSDEstimator.
```
So the doc is generated as
image

Instead of 
```   
"""
       Construct an instance of RankSDEstimator.
```
where we get 
image

File: cryptographic_estimators/RankSDEstimator/ranksd_estimator.py [hunk 10]
```diff
@@ -0,0 +1,145 @@
...
+               +---------------+-------+--------+---------------------------+
+               | algorithm     |  time | memory |         parameters        |
+               +---------------+-------+--------+---------------------------+
+               | ImprovedGRS   | 147.2 |   18.0 |             {}            |
+               | MaxMinors     | 153.0 |   33.0 |     {'a': 12, 'p': 2}     |
+               | SupportMinors | 155.1 |   40.1 | {'b': 1, 'a': 11, 'p': 0} |
+               +---------------+-------+--------+---------------------------+
+
+
+               >>> RSDE = RankSDEstimator(q=2, m=31, n=33, k=15, r=10, w=2, excluded_algorithms=[BasisEnumeration,GRS,OJ1,OJ2,HybridLinearization])
```
Comment (Javierverbel) [inline cryptographic_estimators/RankSDEstimator/ranksd_estimator.py:100]: Is this test already covered by the two above?

**Review by Javierverbel — DISMISSED** [review]
After some minor comments it can be merged from my side.

File: cryptographic_estimators/RankSDEstimator/ranksd_estimator.py (hunk 10 above)
Comment (ricardovp85) [reply cryptographic_estimators/RankSDEstimator/ranksd_estimator.py:100]: It is already kind of covered. As you may note, the difference is that GuessingEnhancedGRS is not excluded. Should I remove it or leave it?

**Review by Javierverbel — APPROVED** [review]

**Review by Memphisd — APPROVED** [review]
looks good!

**Pull request merged**